eclair-core/src/main/scala/fr/acinq/eclair/payment/Bolt12Invoice.scala

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright 2022 ACINQ SAS
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package fr.acinq.eclair.payment
+
+import fr.acinq.bitcoin.Crypto.{PrivateKey, PublicKey}
+import fr.acinq.bitcoin.{Block, ByteVector32, ByteVector64, Crypto}
+import fr.acinq.eclair.crypto.Sphinx.RouteBlinding
+import fr.acinq.eclair.payment.Bolt11Invoice.ExtraHop
+import fr.acinq.eclair.wire.protocol.OfferCodecs.{Bech32WithoutChecksum, invoiceCodec, invoiceTlvCodec}
+import fr.acinq.eclair.wire.protocol.Offers._
+import fr.acinq.eclair.wire.protocol.TlvStream
+import fr.acinq.eclair.{CltvExpiryDelta, Features, InvoiceFeature, MilliSatoshi, MilliSatoshiLong, TimestampSecond}
+import scodec.bits.ByteVector
+
+import java.util.concurrent.TimeUnit
+import scala.concurrent.duration.FiniteDuration
+
+case class Bolt12Invoice(records: TlvStream[InvoiceTlv]) extends Invoice {
+
+  import Bolt12Invoice._
+
+  val amount: MilliSatoshi = records.get[Amount].map(_.amount).get
+
+  override val amount_opt: Option[MilliSatoshi] = Some(amount)
+
+  override val nodeId: Crypto.PublicKey = records.get[NodeId].get.nodeId
+
+  override val paymentHash: ByteVector32 = records.get[PaymentHash].get.hash
+
+  override val paymentSecret: Option[ByteVector32] = None
+
+  override val paymentMetadata: Option[ByteVector] = None
+
+  override val description: Either[String, ByteVector32] = Left(records.get[Description].get.description)
+
+  override val routingInfo: Seq[Seq[ExtraHop]] = Seq.empty
+
+  override val createdAt: TimestampSecond = records.get[CreatedAt].get.timestamp
+
+  override val relativeExpiry: FiniteDuration = FiniteDuration(records.get[RelativeExpiry].map(_.seconds).getOrElse(DEFAULT_EXPIRY_SECONDS), TimeUnit.SECONDS)
+
+  override val minFinalCltvExpiryDelta: Option[CltvExpiryDelta] = records.get[Cltv].map(_.minFinalCltvExpiry)
+
+  override val features: Features[InvoiceFeature] = records.get[FeaturesTlv].map(_.features.invoiceFeatures()).getOrElse(Features.empty)
+
+  override def toString: String = Bech32WithoutChecksum.encode("lni", invoiceCodec, this)
+
+  val chain: ByteVector32 = records.get[Chain].map(_.hash).getOrElse(Block.LivenetGenesisBlock.hash)
+
+  val offerId: Option[ByteVector32] = records.get[OfferId].map(_.offerId)
+
+  private val paths: Option[Seq[RouteBlinding.BlindedRoute]] = records.get[Paths].map(_.paths)
+
+  private val blindedpay: Option[Seq[PayInfo]] = records.get[BlindedPay].map(_.payInfos)
```

Comment:
But do you think it would make more sense as `blindedRoutingInfo` or not? Because we can change the spec name if it helps make the spec easier to understand. I think it makes sense to echo the routing information we provided in Bolt 11.